internal/controller/bucket_controller.go

```diff
@@ -0,0 +1,225 @@
+
+	// Get the bucket that we are reconciling
+	bucket := &v1alpha1.Bucket{}
+	err := r.Client.Get(ctx, req.NamespacedName, bucket)
+	if err != nil {
+		return ctrl.Result{}, errors.WithStack(client.IgnoreNotFound(err))
+	}
+
+	// Create the correct service implementation based on the provider
+	var service objectstorage.ObjectStorageService
+	switch r.ManagementCluster.Provider {
+	case "capa":
+		roleArn, err := r.getRoleArn(ctx)
+		if err != nil {
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+
+		service, err = r.CloudServices.NewS3Service(ctx, r.ManagementCluster.Region, roleArn)
+		if err != nil {
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+	default:
+		return ctrl.Result{}, fmt.Errorf("unsupported provider %s", r.ManagementCluster.Provider)
+	}
+
+	// Handle deleted clusters
+	if !bucket.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, r.reconcileDelete(ctx, service, bucket)
+	}
+
+	// Handle non-deleted clusters
+	return r.reconcileNormal(ctx, service, bucket)
+}
+
+// reconcileCreate creates the s3 bucket.
+func (r *BucketReconciler) reconcileNormal(ctx context.Context, service objectstorage.ObjectStorageService, bucket *v1alpha1.Bucket) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	originalBucket := bucket.DeepCopy()
+	// If the Bucket doesn't have our finalizer, add it.
+	if controllerutil.AddFinalizer(bucket, v1alpha1.BucketFinalizer) {
+		// Register the finalizer immediately to avoid orphaning AWS resources on delete
+		if err := r.Client.Patch(ctx, bucket, client.MergeFrom(originalBucket)); err != nil {
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+	}
+
+	logger.Info("Checking if bucket exists")
+	exists, err := service.ExistsBucket(ctx, bucket)
+	if err != nil {
+		logger.Info(fmt.Sprintf("Either you don't have access to bucket %v or another error occurred. "+
+			"Here's what happened: %v", bucket.Spec.Name, err))
+		return ctrl.Result{}, errors.WithStack(err)
+	} else if !exists {
+		logger.Info(fmt.Sprintf("Bucket %v is available, creating", bucket.Spec.Name))
+		err = service.CreateBucket(ctx, bucket)
+		if err != nil {
+			logger.Error(err, "Bucket could not be created")
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+	} else {
+		logger.Info(fmt.Sprintf("Bucket %v exists and you already own it.", bucket.Spec.Name))
+	}
+
+	logger.Info("Configuring bucket settings")
+	// If expiration is not set, we remove all lifecycle rules
+	err = service.ConfigureBucket(ctx, bucket)
+	if err != nil {
+		logger.Error(err, "Bucket could not be configured")
+		return ctrl.Result{}, errors.WithStack(err)
+	}
+
+	logger.Info("Bucket ready")
+	originalBucket = bucket.DeepCopy()
+	bucket.Status = v1alpha1.BucketStatus{
+		BucketID:    bucket.Spec.Name,
+		BucketReady: true,
+	}
+	if err := r.Client.Status().Patch(ctx, bucket, client.MergeFrom(originalBucket)); err != nil {
+		return ctrl.Result{}, errors.WithStack(client.IgnoreNotFound(err))
```

Comment:
```suggestion
		return ctrl.Result{}, errors.WithStack(err)
```

If it doesn't exist here, then something wrong happened and we want to know about it :smile:

Reply:
I think I added it because the test remove it before this is run sometimes. Let me try to fix it

Reply:
I confirm I have this:

```
[FAILED] in [It] - /home/quentin/Documents/code/object-storage-operator/internal/controller/bucket_controller_test.go:138 @ 10/30/23 13:57:34.565
  << Timeline

  [FAILED] Unexpected error:
      <*errors.withStack | 0xc0006aceb8>: 
      buckets.objectstorage.giantswarm.io "my-bucket-name" not found
      {
          error: <*errors.StatusError | 0xc00028b540>{
              ErrStatus: {
                  TypeMeta: {Kind: "", APIVersion: ""},
```

Reply:
I'm not really sure why the patch would say the bucket is missing 🤔